src/transformers/models/seamless_m4t/modeling_seamless_m4t.py

```diff
@@ -0,0 +1,3944 @@
+            past_key_values=past_key_values,
+            inputs_embeds=decoder_inputs_embeds,
+            use_cache=use_cache,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+        )
+
+        lm_logits = self.lm_head(decoder_outputs.last_hidden_state) + self.final_logits_bias
+
+        masked_lm_loss = None
+        if labels is not None:
+            loss_fct = CrossEntropyLoss()
+            masked_lm_loss = loss_fct(lm_logits.view(-1, self.config.vocab_size), labels.view(-1))
+
+        if not return_dict:
+            outputs = decoder_outputs + encoder_outputs
+            output = (lm_logits,) + outputs[1:]
+            return ((masked_lm_loss,) + output) if masked_lm_loss is not None else output
+
+        return Seq2SeqLMOutput(
+            loss=masked_lm_loss,
+            logits=lm_logits,
+            past_key_values=decoder_outputs.past_key_values,
+            decoder_hidden_states=decoder_outputs.hidden_states,
+            decoder_attentions=decoder_outputs.attentions,
+            cross_attentions=decoder_outputs.cross_attentions,
+            encoder_last_hidden_state=encoder_outputs.last_hidden_state,
+            encoder_hidden_states=encoder_outputs.hidden_states,
+            encoder_attentions=encoder_outputs.attentions,
+        )
+
+    def prepare_inputs_for_generation(
+        self,
+        decoder_input_ids,
+        past_key_values=None,
+        attention_mask=None,
+        head_mask=None,
+        decoder_head_mask=None,
+        cross_attn_head_mask=None,
+        use_cache=None,
+        encoder_outputs=None,
+        **kwargs,
+    ):
+        # cut decoder_input_ids if past is used
+        if past_key_values is not None:
+            decoder_input_ids = decoder_input_ids[:, -1:]
+
+        return {
+            "input_ids": None,  # encoder_outputs is defined. input_ids not needed
+            "encoder_outputs": encoder_outputs,
+            "past_key_values": past_key_values,
+            "decoder_input_ids": decoder_input_ids,
+            "attention_mask": attention_mask,
+            "head_mask": head_mask,
+            "decoder_head_mask": decoder_head_mask,
+            "cross_attn_head_mask": cross_attn_head_mask,
+            "use_cache": use_cache,
+        }
+
+    @staticmethod
+    def _reorder_cache(past_key_values, beam_idx):
+        reordered_past = ()
+        for layer_past in past_key_values:
+            # cached cross_attention states don't have to be reordered -> they are always the same
+            reordered_past += (
+                tuple(past_state.index_select(0, beam_idx) for past_state in layer_past[:2]) + layer_past[2:],
+            )
+        return reordered_past
+
+
+@add_start_docstrings(
+    "The text-to-speech SeamlessM4T Model transformer which can be used for T2ST.",
+    SEAMLESS_M4T_START_DOCSTRING,
+)
+class SeamlessM4TForTextToSpeech(SeamlessM4TForTextToText):
```

Comment:
Let's not use a `super()` call -> we should just define the `SeamlessM4TForTextToText` as an attribute of `SeamlessM4TForTextToSpeech` in the `__init__`:
```python
self.text_to_text = SeamlessM4TForTextToText(config)
```

Reply:
I'd rather keep it as it is for one main reason - it allows to seamlessly load `SeamlessM4TForTextToText`, `SeamlessM4TForTextToSpeech`, `SeamlessM4TForSpeechToText` and `SeamlessM4TForSpeechToSpeech` from the same checkpoint saved by `SeamlessM4TModel`.

If I change the architecture of `SeamlessM4TForTextToSpeech` with your proposal, I would no longer have one shared architecture naming across the 5 models, since it doesn't make sense to have `self.text_to_text` and `self.speech_to_text` in `SeamlessM4TModel` because of duplication of the `text_decoder`. 

WDYT ? Do you agree?

Reply:
We're pretty anti using `super()` calls in the forward method of `transformers` models. You'll see in the rest of the library that it's always preferred to write out the models explicitly using their sub-components, since this helps the reader i) understand the architecture and ii) track the modelling calls.

I don't fully understand why the weight loading would not work if we explicitly defined the sub-models here? Regarding having the same state dict, setting:
```python
class SeamlessM4TForTextToSpeech(SeamlessM4TForTextToText):
   def __init__(self, config: SeamlessM4TConfig):
        super().__init__(config)

        self.t2u_model = SeamlessM4TTextToUnitForConditionalGeneration(config)
        self.vocoder = SeamlessM4TCodeHifiGan(config)
```
Is the same as:
```
class SeamlessM4TForTextToSpeech(SeamlessM4TPreTrainedModel):
   def __init__(self, config: SeamlessM4TConfig):
        super().__init__(config)
        # text-to-text: these three modules are equivalent to `SeamlessM4TForTextToText`
        self.text_encoder = SeamlessM4TEncoder(config)
        self.text_decoder = SeamlessM4TDecoder(config)
        self.lm_head = nn.Linear(config.hidden_size, config.vocab_size, bias=False)

        # text-to-units
        self.t2u_model = SeamlessM4TTextToUnitForConditionalGeneration(config)

        # units-to-waveform
        self.vocoder = SeamlessM4TCodeHifiGan(config)
```
=> can we not just write them out explicitly like this? As you have done for the `SeamlessM4TModel` for example? 
https://github.com/huggingface/transformers/blob/ec4b204abf8d2060d3687b89045ad783befba273/src/transformers/models/seamless_m4t/modeling_seamless_m4t.py#L3558

Reply:
Ah yes, in that case it makes sense indeed! I wanted to avoid that because the code is already really big but if it's clearer, I'll do as suggested, thanks for the suggestion!

There will still be some `super().generate()`, but there are no other choices here in terms of design.